src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -341,15 +343,48 @@ void primitive_inst::update_impl() {
     };
 
     if (!_node->is_type<data>() && !(_node->is_type<mutable_data>() && _node->get_dependencies().empty())) {
+        GPU_DEBUG_GET_INSTANCE(debug_config);
         // Update param if fake_alignment is available
         auto updated_params = _node->type()->get_fake_aligned_params(*_impl_params);
         auto layout_key = get_layout_key(updated_params);
         auto& cache = get_network().get_implementations_cache();
-        if (cache.has(layout_key)) {
-            _impl = cache.get(layout_key)->clone();
-            GPU_DEBUG_PROFILED_STAGE_CACHE_HIT(true);
-        } else {
+        bool has_cached_impl = false;
+        {
+            std::lock_guard<std::mutex> lock(get_network().get_impl_cache_mutex());
```

Comment:
Why does this need lock? Currently impl_cache is owned by each network.

Reply:
Self answer: Seems that the lock is for the case when the same layout is already built in the compilation context.
Now my thought is that, if we cannot avoid this lock, then it would be better to  move the impl cache to program, for better cache hit ratio & memory usage reduction.

Reply:
@ahnyoung-paul  FYI